### PR TITLE
Feature/component fast clock

### PR DESCRIPTION
**Pull request by jmvdwerf**
The fast clock component provides the fast_clock functionality to give loconet components an internal fast clock that reacts on the global fast clock messages, and in between keeps a clock ticking according to the global loconet clock settings. After each received fast clock message, the internal structure is updated.

### CODE REVIEW

File: src/components/fast_clock.c [hunk 1]
```diff
@@ -0,0 +1,253 @@
+/*
+ * @file fast_clock.c
+ * @brief Loconet clock implementation
+ *
+ * \copyright Copyright 2017 /Dev. All rights reserved.
+ * \license This project is released under MIT license.
+ *
+ * @author Jan Martijn van der Werf
+ * @see fast_clock.h
```
Comment (fuegas) [inline src/components/fast_clock.c]: This line is obsolete.

File: src/components/fast_clock.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
+/*
+ * @file fast_clock.c
+ * @brief Loconet clock implementation
```
Comment (fuegas) [inline src/components/fast_clock.c]: `clock` or `fast clock`?

File: src/components/fast_clock.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+ *
+ * @author Jan Martijn van der Werf
+ * @see fast_clock.h
+ */
+
+#include "fast_clock.h"
+
+// ----------------------------------------------------------------------------
+// Prototypes
+void fast_clock_handle_update_(FAST_CLOCK_TIME_Type time);
```
Comment (fuegas) [inline src/components/fast_clock.c]: If this is a dummy function, please let it end with `_dummy` as other functions do.

File: src/components/fast_clock.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+// ----------------------------------------------------------------------------
+// this is the function that any system that wants to react on the clock should
```
Comment (fuegas) [inline src/components/fast_clock.c]: Please start sentences with a Capital. For all comments 👍

File: src/components/fast_clock.c [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  uint8_t millisecond;
+} FAST_CLOCK_STATUS_Type;
+
+FAST_CLOCK_STATUS_Type fast_clock_status = {0, 0, 0, 0, 1, 0};
+
+uint16_t fast_clock_current_intermessage_delay = 0;
+
+// ----------------------------------------------------------------------------
+// The clock is defined on div8. That means that the clock is triggered
+// every 50ms.
```
Comment (fuegas) [inline src/components/fast_clock.c]: The clock is triggered faster, every 1us. But with the delay defined below you get to 50ms. Please reflect that in the comment.

File: src/components/fast_clock.c [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+  fast_clock_timer->COUNT16.INTENSET.reg = TC_INTENSET_MC(1);
+  NVIC_EnableIRQ(nvic_irqn);
+
+  fast_clock_timer->COUNT16.COUNT.reg = 0;
+  fast_clock_timer->COUNT16.CC[0].reg = FAST_CLOCK_TIMER_DELAY;
+  fast_clock_timer->COUNT16.CTRLA.reg |= TC_CTRLA_ENABLE;
+}
+
+// ----------------------------------------------------------------------------
+// sets the values for being a master: whether it is a
```
Comment (fuegas) [inline src/components/fast_clock.c]: This reads like an unfinished comment. Furthermore, I'd like to suggest to convert the function to `fast_clock_set_master` (and subsequent `fast_clock_set_slave` a little further down).

File: src/components/fast_clock.c [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+//----------------------------------------------------------------------------
+// this function processes the clock information sent on the loconet
+void loconet_rx_fast_clock(uint8_t *data, uint8_t length)
+{
+  if (length < 8)
+  {
+    return;
+  }
+
+  if (!(data[7] == 1)) {
```
Comment (fuegas) [inline src/components/fast_clock.c]: `if (!(data[7] == 1)) {` can be written as `if (data[7] != 1) {`.

File: src/components/fast_clock.c [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    // the message is not a correct clock tick!
+    return;
+  }
+
+  //1st byte of data is the clock rate
```
Comment (fuegas) [inline src/components/fast_clock.c]: Missing space after `//`.

File: src/components/fast_clock.c [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+  fast_clock_set_rate(data[0]);
+
+  current_time.minute = data[3] - (128-60);
+  current_time.hour = data[5] >= (128-24) ? data[5] - (128-24) : data[5] % 24;
```
Comment (fuegas) [inline src/components/fast_clock.c]: You use spaces around `>=`, `?`, `:` and `%`, why not around `-`?

File: src/components/fast_clock.c [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  current_time.day = data[6] % 7;
+  current_time.second = 0;
```
Comment (fuegas) [inline src/components/fast_clock.c]: Above you have the flow `minute`, `hour`, `day` and then `second`. To keep the flow, you can place `second` before `minute`. Same for the `millisecond` below.

File: src/components/fast_clock.c [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    current_time.hour,
+    current_time.day,
+    fast_clock_status.id1,
+    fast_clock_status.id2
+  );
+}
+
+
+// ----------------------------------------------------------------------------
+// this function should be called every 50ms. It updates the clock with
```
Comment (fuegas) [inline src/components/fast_clock.c]: `should be called every 50ms`, or `is called every 50ms`?

File: src/components/fast_clock.c [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+void fast_clock_tick(void)
+{
+  // keeps track whether we need an update. We only send an update
+  // if the minute counter has been increased.
+  bool notify = false;
+
+  // update the milliseconds passed with the clock rate
+  fast_clock_status.millisecond += fast_clock_status.rate;
+
+  // as the millisecond counter counts every 50ms, we have 200 cycles for a
```
Comment (fuegas) [inline src/components/fast_clock.c]: `we have 200 cycles` or `we need to count to 200`?

File: src/components/fast_clock.c [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    current_time.second = 0;
+    notify = true;
+  }
+
+  if (current_time.minute > 59)
+  {
+    current_time.minute = 0;
+    current_time.hour++;
+  }
+  if (current_time.hour > 23)
```
Comment (fuegas) [inline src/components/fast_clock.c]: Between second and minute you have a newline, between minute and hour not. Please make it consistent, either way 👍

File: src/components/fast_clock.h [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+ * id1 and id2 are used to identify the clock, the intermessage_delay
+ * states the seconds between every two clock messages.
+ *
+ * To return to slave mode, use
+ *
+ *     fast_clock_set_as_slave();
+ *
+ * To use the clock system, one should initialize a clock using
+ *
+ *    LOCONET_BUILD(timer)
```
Comment (fuegas) [inline src/components/fast_clock.h]: `LOCONET_BUILD`? I think it should be `FAST_CLOCK_BUILD`?

File: src/components/fast_clock.h [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+ * It reacts on the fast clock messages of loconet to sync the clock.
+ * Internally, it updates the clock, until a new fast clock message
+ * arrives, after which it resets the clock to the received message,
+ * and starts ticking again, using the appropriate clock rate.
+ *
+ * Additionally, the system can run as a clock master, i.e., it sends
+ * the fast clock messages itself. For this, set the appropriate values
+ * using:
+ *
+ *    fast_clock_set_as_master(id1, id2, intermessage_delay)
```
Comment (fuegas) [inline src/components/fast_clock.h]: If the name of the function is changed, it should be done here as well. Same for slave below.

File: src/components/fast_clock.h [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+#define _LOCONET_FAST_CLOCK_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include "loconet/loconet_tx_messages.h"
+#include "utils/logger.h"
+
+typedef struct {
+  uint8_t minute;
+  uint8_t second;
```
Comment (fuegas) [inline src/components/fast_clock.h]: Please order the variables 👍

File: src/components/fast_clock.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+// the system acts as a clock master.
+extern void fast_clock_set_rate(uint8_t);
+
+// ------------------------------------------------------------------
+// Sets the time
+extern void fast_clock_set_time(FAST_CLOCK_TIME_Type time);
+
+// ------------------------------------------------------------------
+// Returns the minutes
+extern uint8_t fast_clock_get_minutes(void);
```
Comment (fuegas) [inline src/components/fast_clock.h]: I would remove `get_{minutes,hours,day}` in favor of `get_time`.

File: src/loconet/loconet_tx_messages.h [hunk 18]
```diff
@@ -30,4 +30,9 @@ extern void loconet_tx_sw_rep(uint16_t address, bool state);           // 0xB1
 extern void loconet_tx_input_rep(uint16_t address, bool state);        // 0xB2
 extern void loconet_tx_long_ack(uint8_t lopc, uint8_t ack1);           // 0xB4
 
+
```
Comment (fuegas) [inline src/loconet/loconet_tx_messages.h]: One newline above and below would match previous blocks.

**Review by fuegas — CHANGES_REQUESTED** [review]
My biggest concern is dat the interrupt that is handled takes a lot of time in which other interrupts won't be handled. 

My suggestion is to make the `tick` function as small as can be (like, only add the rate to the milliseconds)  and add a function to process fast clock internals in the main loop.

File: src/components/fast_clock.c (hunk 2 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Check

File: src/components/fast_clock.c (hunk 1 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Removed

File: src/components/fast_clock.c (hunk 3 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Check

File: src/components/fast_clock.c (hunk 5 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: fixed

File: src/components/fast_clock.c (hunk 6 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Done!

File: src/components/fast_clock.c (hunk 7 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Check!

File: src/components/fast_clock.c (hunk 9 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Hmm, good question...

File: src/components/fast_clock.c (hunk 13 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Check!

File: src/components/fast_clock.c (hunk 12 above)
Comment (jmvdwerf) [reply src/components/fast_clock.c]: Explained in comments. In fact I should add 50 * rate to the millisec counter, and then check whether this is >= 1000. Just dividing the lot by 50 is simpler :)

File: src/components/fast_clock.h (hunk 15 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: Check!

File: src/loconet/loconet_tx_messages.h (hunk 18 above)
Comment (jmvdwerf) [reply src/loconet/loconet_tx_messages.h]: Removed

File: src/components/fast_clock.h (hunk 17 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: Removed

**Comment by jmvdwerf** [comment]
Comments addressed!

File: src/components/fast_clock.h (hunk 15 above)
Comment (fuegas) [reply src/components/fast_clock.h]: You say check, but I do not see a change here 😉

File: src/components/fast_clock.h [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+ * using:
+ *
+ *    fast_clock_set_as_master(id1, id2, intermessage_delay)
+ *
+ * id1 and id2 are used to identify the clock, the intermessage_delay
+ * states the seconds between every two clock messages.
+ *
+ * To return to slave mode, use
+ *
+ *     fast_clock_set_as_slave();
```
Comment (fuegas) [inline src/components/fast_clock.h]: Neither here.

File: src/components/fast_clock.h [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+// Returns the current time as FAST_CLOCK_TIME_Type
+extern FAST_CLOCK_TIME_Type fast_clock_get_time(void);
+
+// ------------------------------------------------------------------
+// Returns the time as hour * 100 + minutes
+extern uint16_t fast_clock_get_time_as_int(void);
+
+// ------------------------------------------------------------------
+// This is the IRQ function that is called after every clock cycle.
+extern void fast_clock_tick(void);
```
Comment (fuegas) [inline src/components/fast_clock.h]: If it is an IRQ function, I'm missing IRQ in the name.

**Review by fuegas — CHANGES_REQUESTED** [review]
The inline comments are processed correctly (almost all). But the main concern I noted was not handled:

--8<--
My biggest concern is dat the interrupt that is handled takes a lot of time in which other interrupts won't be handled.

My suggestion is to make the tick function as small as can be (like, only add the rate to the milliseconds) and add a function to process fast clock internals in the main loop.
--8<--

File: src/components/fast_clock.h (hunk 15 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: I missed them in the comments... Comments updated...

File: src/components/fast_clock.h (hunk 19 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: I missed them in the comments... Comments updated...

File: src/components/fast_clock.h (hunk 14 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: fixed

File: src/components/fast_clock.h (hunk 16 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: did that

**Comment by jmvdwerf** [comment]
I changed it to a fast_clock_loop function, but I find it ugly...

**Review by fuegas — APPROVED** [review]

File: src/components/fast_clock.h (hunk 20 above)
Comment (jmvdwerf) [reply src/components/fast_clock.h]: Renamed

**Pull request merged**